liblwgeom/lwrandom.c

```diff
@@ -0,0 +1,278 @@
+/**********************************************************************
+ *
+ * PostGIS - Spatial Types for PostgreSQL
+ * http://postgis.net
+ *
+ * PostGIS is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * PostGIS is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with PostGIS.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ **********************************************************************
+ *
+ * Copyright 2019 Mike Taves
+ *
+ **********************************************************************/
+
+
+#include <stdlib.h>
+#include <stdint.h>
+/* #include <stdio.h> */
+
+#include "lwrandom.h"
+#include <stdlib.h>
+#include <time.h>
+
+
+static unsigned char _lwrandom_seed_set = 0;
+static int32_t _lwrandom_seed[3] = {0x330e, 0xabcd, 0x1234};
+
+/*
+ * Set seed for a random number generator.
+ * A zero value uses clock as seed the first time only.
+ */
+void
+lwrandom_set_seed(uint32_t seed)
+{
+	if (seed == 0)
+	{
+		if (_lwrandom_seed_set == 0)
+			seed = ((unsigned int)time(NULL) * 1996) << 8;
+		else
+			return;
+	}
+
+	_lwrandom_seed[1] = (seed % 2147483562) + 1; /* value between 1 and 2147483562 */
+	_lwrandom_seed[2] = (((seed + 6) >> 12) % 2147483398) + 1; /* value between 1 and 2147483398 */
+	_lwrandom_seed_set = 1;
+}
+
+/*
+ * Generate a random floating-point value.
+ * Values are uniformly distributed between 0 and 1.
+ *
+ * Author:
+ *   Original PASCAL version by Pierre L'Ecuyer.
+ *   C version by John Burkardt, modified by Mike Taves.
+ *
+ * Reference:
+ *   Pierre L'Ecuyer,
+ *   Efficient and Portable Combined Random Number Generators,
+ *   Communications of the ACM,
+ *   Volume 31, Number 6, June 1988, pages 742-751.
+ */
+double
```

Comment:
This function comes from an external project. It seems like it's under GPL3, so we'd need to check if it's compatible with Postgis' GPL2 and  verify if it needs to be provided along with the source.

Reply:
The function was modified from [asa183.c](https://people.sc.fsu.edu/~jburkardt/c_src/asa183/asa183.c) (look for `double r8_uni ( int *s1, int *s2 )`). There are about 12 lines of effective code used here, most of which I've modified to suite this purpose and coding style (e.g. tabs). I'm not sure how different this qualifies to say it's an entirely "different" function with a different name, calling parameters and data types. As for downgrading LGPL code from v. 3 to 2, the answer appears to be [no](https://softwareengineering.stackexchange.com/q/167457).

The primary source for this algorithm, and I've checked, is Figure 3 "A Portable Generator for 32-bit Computers" from Pierre L'Ecuyer's [1988 manuscript](https://doi.org/10.1145/62959.62969). While it is in Pascal, the algorithm is identical, with about 12 effective statements to make it operate.

If using John Burkardt's 12 lines of C code is an issue, I can remove credit from John Burkardt, and just cite Pierre L'Ecuyer's 1988 work, as there are not too many different ways to rewrite a few lines of Pascal into C. Or, find a different algorithm (there are many PRNGs out there). I've selected this algorithm due to it's simplicity and apparent portability on 32-bit integers.